include/swift/AST/PackExpansionMatcher.h

```diff
@@ -63,49 +63,78 @@ class TuplePackMatcher {
   bool match();
 };
 
-/// Performs a structural match of two lists of (unlabeled) function
-/// parameters.
+/// Performs a structural match of two lists of types.
 ///
 /// The invariant is that each list must only contain at most one pack
 /// expansion type. After collecting a common prefix and suffix, the
 /// pack expansion on either side asborbs the remaining elements on the
 /// other side.
-class ParamPackMatcher {
-  ArrayRef<AnyFunctionType::Param> lhsParams;
-  ArrayRef<AnyFunctionType::Param> rhsParams;
+class TypeListPackMatcher {
+  struct Element {
```

Comment:
Could we avoid conversion into the common Element representation by making the TypeListPackMatcher a template parameterized by the element type?

Reply:
Ah I think I know what you mean, we could parametrize over Element and then lazily convert while matching, I'll give it a try.